File: spotify.py
import requests 
import json 
import math
import os 
import random 
import time 
from itunes import itunes,UnFoundException
import json 

class spotify():
    def __init__(self):
        with open("keys.json") as f:
            self.data = json.load(f)            
        self.redirect_uri = "http://localhost:8888/callback"
        self.scopes = "user-library-read user-read-playback-state"
        self.headers = {"Authorization": "Bearer " +  self.data['access_token'],"Content-Type": "application/json "}
        authorize_url = f"https://accounts.spotify.com/authorize?client_id={self.data['client_id']}&response_type=code&redirect_uri={self.redirect_uri}&scope={self.scopes.replace(' ','%20')}"
        self.start = time.time()
        self.liked_albums = []
        if self.data['refresh_token'] == "":
            code = input(f"Click here and copy paste your authorization code {authorize_url}:  ")
            self.get_refresh(code)
        
        self.get_access()



    def get_access(self):
        r = requests.post(f"https://accounts.spotify.com/api/token?grant_type=refresh_token&refresh_token={self.data['refresh_token']}&client_id={self.data['client_id']}&client_secret={self.data['client_secret']}",headers={'Content-Type': 'application/x-www-form-urlencoded'})
        if r.status_code == 200:
            print("successfully refreshed access token")
            self.data['access_token'] = r.json()['access_token']
            self.headers = {"Authorization": "Bearer " +  self.data['access_token'],"Content-Type": "application/json "}

            with open("keys.json","w") as f:        
                json.dump(self.data,indent=2,fp=f)

        with open("keys.json") as f:
            self.data = json.load(f)


    def get_refresh(self,code):
        r = requests.post(f"https://accounts.spotify.com/api/token?grant_type=authorization_code&code={code}&redirect_uri={self.redirect_uri}&client_id={self.data['client_id']}&client_secret={self.data['client_secret']}",headers={'Content-Type': 'application/x-www-form-urlencoded'})
        if r.status_code == 200:
            print("successfully got the refresh token")
            self.data['refresh_token'] = r.json()['refresh_token']
            with open("keys.json","w") as f:        
                json.dump(self.data,indent=2,fp=f)



    def album_images(self):  # gets unique album images from your liked songs 
        albums = []
        s = requests.Session()
        s.headers.update(self.headers)
    
        r = s.get(f"https://api.spotify.com/v1/me/tracks?limit=50&offset=0")
        if r.status_code == 200:
            offset = r.json()['total']
            for i in range(math.ceil(offset / 50)): 
                r = s.get(f"https://api.spotify.com/v1/me/tracks?limit=50&offset={i * 50}")
                for items in r.json()['items']:
                    album_image = items['track']['album']['images'][0]['url']
                    album_name = items['track']['album']['name']
                    album_id = items['track']['album']['id']
                    if album_name not in albums:
                        albums.append({"name":album_name,"image":album_image,"id":album_id})     
            return albums

        if r.status_code == 401:
            self.get_access()
        else:
            return False 
        # potential data to show: album information e.g. name, date, album count 

    def album_lookup(self,album_id): # finds tracklist and information about the specified album  (limit to 50 songs in a playlist atm, need to change for huge albums)
        r = requests.get(f"https://api.spotify.com/v1/albums/{album_id}",headers=self.headers)
        if r.status_code == 200:
            data = r.json()
            album_length = 0 
            tracks = []            
            for track in data['tracks']['items']:
                tracks.append({"name":track['name'],"position":track['track_number'],"duration":track['duration_ms']})
                
            for t in tracks:
                album_length += t['duration']
            return {"album_type":data['album_type'],"artist":data['artists'],"record-label":data['label'],"release-date":data['release_date'],"tracks":tracks,"disc_amount":data['tracks']['items'][-1]['disc_number'],"album_duration_ms":album_length}
    
    def playback_state(self):
        r = requests.get("https://api.spotify.com/v1/me/player",headers=self.headers)
        if r.status_code == 200:
            return r.json()
        elif r.status_code == 401:
            self.get_access()
            return {"is_playing":False} 
        else:
            return {"is_playing":False}  
        

    def info(self):
        images_downloaded = os.listdir("./images/compressed/")
        total_info = {}
        playback = self.playback_state()
        if 'item' in playback:
            if playback['item'] == None: 
                return False 
            
        if playback['is_playing'] == True and playback['item']['is_local'] != True: # ignores currently playing local file because i want to display album image which isnt provided
            print("listening to music")
            item = playback['item']
            album = item['album']
            album_name = album['name']
            image = album['images'][0]['url']
            album_id = item['album']['id']
            listening_to = item['name']   
            progress = playback['progress_ms']
            device_type = item['type']

            total_info['currently_playing'] =  {"listening_to":listening_to,"progress_ms":progress,"device_type":device_type}
        else:
            print("not listening to music")
            if self.liked_albums == []:
                self.liked_albums = self.album_images()
            
            album = random.choice(self.liked_albums)
            album_name = album['name']
            image = album['image']
            album_id = album['id']


        lookup_album = self.album_lookup(album_id)
        total_info['album_id'] =  album_id
        total_info['album_info'] = lookup_album 
        total_info['album_info']['album'] = album_name
        total_info['album_info']['image'] = image

        if f"{album_id}.jpg" not in images_downloaded: 
            img_data = requests.get(image).content
            with open(f'./images/compressed/{album_id}.jpg', 'wb') as f:
                print(f"downloading: {album_id}")
                f.write(img_data) 

        return total_info

File: spotify.cpp
/*
    spotify api used to get liked songs and other info like the songs you are currently listening 

    . follow auth flow (getting access token and refresh using client secret/id)
        refresh
        https://accounts.spotify.com/api/token?grant_type=refresh_token&refresh_token={refresh_token}&client_id={client_id}&client_secret={client_secret}
        
        access
        https://accounts.spotify.com/api/token?grant_type=authorization_code&code={code}&redirect_uri={redirect_uri}&client_id={client_id}&client_secret={client_secret}
    
    
    . get liked albums and order them all 

    . dont really need to get other spotify data due to itunes album giving this 

    https://api.spotify.com/v1/me/player

    https://api.spotify.com/v1/me/tracks?limit=60&offset=0

    scopes: "user-library-read user-read-playback-state"
    headers: {"Authorization": "Bearer {access_token]}","Content-Type": "application/json "}
*/
#include "header.h"
#include <iostream>
#include <fstream>
#include <nlohmann/json.hpp>
#include <variant>  

// get the token wanted from json file
std::string get_token(std::string token_type){
    std::ifstream file("tokens.json");
    nlohmann::json data;
    file >> data;
    file.close();
    std::string token_wanted = data[token_type];
    return token_wanted;
}

// replace value of key from json file 
int write_token(std::string token_type, std::string value_wanted){
    std::ifstream file("tokens.json");
    nlohmann::json data;
    file >> data;
    file.close();

    data[token_type] = value_wanted;


    std::ofstream f("tokens.json");
    f << data.dump(4); // Use dump function to format with indentation
    f.close();
    return 0;
} 


std::string redirect_uri = "http://localhost:8888/callback";

// get CODE= value from giving application permissions, will send you to the spotify site to give permission if you already gave permission it will just go straight to the redirect uri  
void first(){
    std::string scopes = replace_spaces("user-library-read user-read-playback-state");
    std::cout <<  "https://accounts.spotify.com/authorize?client_id=" << get_token("client_id")  << "&response_type=code&redirect_uri=" << redirect_uri << "&scope=" <<  scopes << "\n";
}

// use code that was received after giving permission to app to get refresh token 
void get_tokens(std::string code){
    std::stringstream params;
    params<< "grant_type=authorization_code&code=" << code << "&redirect_uri=" << redirect_uri << "&client_id=" << get_token("client_id") << "&client_secret="<< get_token("client_secret");
    nlohmann::json result =  post_request("https://accounts.spotify.com/api/token",params.str());
    if (result.find("error") == result.end()){ // if error isnt found 
        // {access_token:"",refresh_token:""}
        write_token("access_token",result["access_token"]);
        write_token("refresh_token",result["refresh_token"]);
    }
}

// use refresh token to get new access token 
void refresh_access(){
    std::stringstream params; 
    params << "grant_type=refresh_token&refresh_token=" << get_token("refresh_token") << "&client_id=" << get_token("client_id") << "&client_secret=" << get_token("client_secret");
    nlohmann::json result =  post_request("https://accounts.spotify.com/api/token",params.str());
    if (result.find("error") == result.end()){
        std::cout << result << "\n";           
        write_token("access_token",result["access_token"]);
    }
}


// playback state 
void playback_state(){    
    nlohmann::json results = get_request("https://api.spotify.com/v1/me/player",true);
    std::cout << results << "\n";
}


//get all liked songs
std::vector<nlohmann::json> get_liked_songs(){
    // need to paginate to end of results 
    int page = 0;
    bool next_page = true;
    std::vector<nlohmann::json> data; //[{items:[1,2,3]},{items:[1,2,3]}]
    while (next_page  == true){
        nlohmann::json results = get_request("https://api.spotify.com/v1/me/tracks?limit=50&offset=" + std::to_string(page * 50),true);
        std::cout << "offset: " << results["offset"] << " total: " << results["total"]  << "\n";
        //std::cout << results["items"].back() << "\n"
        if (results["items"].size() > 0){
            data.insert(data.end(), results);
        }

        if (results["offset"] > results["total"]){
            next_page = false; 
        }   
        
        ++page;
    }
    return data;

}

bool is_in_vec_dics(std::vector<std::unordered_map<std::string, std::string>> list_of_dicts,std::string key, std::string value){
    for (auto d: list_of_dicts){
        if (d[key] == value){
            return true;
        }

    }
    return false;
}

//organises them to unique albums
std::vector<std::unordered_map<std::string, std::string>> get_unique_albums(){
    std::vector<std::unordered_map<std::string, std::string>> unique_albums;
    std::vector<nlohmann::json> liked = get_liked_songs();
    for (auto page : liked){
        for (auto data : page["items"]){
            std::string album_name = data["track"]["album"]["name"];
            std::string album_url = data["track"]["album"]["images"][0]["url"];
            std::string artist =  data["track"]["album"]["artists"][0]["name"];
            std::string album_type = data["track"]["album"]["album_type"];
            if (is_in_vec_dics(unique_albums,"name",album_name) == false){
                // name is unique
                std::unordered_map<std::string, std::string> album;
                album["album_type"] = album_type;
                album["name"] = album_name;
                album["image"] = album_url;
                album["artist"] = artist;
                
                unique_albums.push_back(album);
            }

        
        }
    }



    return unique_albums;
} 








